File: scripts/newsletter/fetch-via-defuddle.js
#!/usr/bin/env node
// mt-webfetch fallback fetcher via defuddle.md
// Usage: node fetch.js <target_url>
// Exit codes: 0 = content returned, 1 = empty/failed, 2 = bad args

const url = process.argv[2];
if (!url) {
  console.error("Usage: node fetch.js <target_url>");
  process.exit(2);
}

const defuddleUrl = `https://defuddle.md/${url}`;

async function main() {
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), 30000);
  try {
    const res = await fetch(defuddleUrl, {
      redirect: "follow",
      signal: controller.signal,
      headers: { "User-Agent": "mt-webfetch/1.0" },
    });
    clearTimeout(timeout);
    const body = await res.text();
    if (!res.ok || !body || body.trim().length === 0) {
      console.error(`mt-webfetch: defuddle returned ${res.status} / empty body for ${url}`);
      process.exit(1);
    }
    process.stdout.write(body);
  } catch (err) {
    clearTimeout(timeout);
    console.error(`mt-webfetch: fetch failed for ${url}: ${err.message}`);
    process.exit(1);
  }
}

main();

File: scripts/newsletter/fetch_via_defuddle.go
// mt-webfetch fallback fetcher via defuddle.md.
// Usage: go run ./scripts/newsletter fetch-via-defuddle <target_url>
// Exit codes: 0 = content returned, 1 = empty/failed, 2 = bad args.
// Note: `go run` collapses any nonzero program exit to 1, so callers invoking
// through `go run` can only distinguish success from failure — the stderr
// diagnostics carry the detail.
package main

import (
	"fmt"
	"io"
	"net/http"
	"os"
	"strings"
	"time"
)

func runFetchViaDefuddle(args []string) {
	if len(args) < 1 || args[0] == "" {
		fmt.Fprintln(os.Stderr, "Usage: go run ./scripts/newsletter fetch-via-defuddle <target_url>")
		os.Exit(2)
	}
	target := args[0]
	defuddleURL := "https://defuddle.md/" + target

	fail := func(err error) {
		fmt.Fprintf(os.Stderr, "mt-webfetch: fetch failed for %s: %v\n", target, err)
		os.Exit(1)
	}

	req, err := http.NewRequest(http.MethodGet, defuddleURL, nil)
	if err != nil {
		fail(err)
	}
	req.Header.Set("User-Agent", "mt-webfetch/1.0")
	res, err := (&http.Client{Timeout: 30 * time.Second}).Do(req)
	if err != nil {
		fail(err)
	}
	defer res.Body.Close()
	body, err := io.ReadAll(res.Body)
	if err != nil {
		fail(err)
	}
	ok := res.StatusCode >= 200 && res.StatusCode < 300
	if !ok || strings.TrimSpace(string(body)) == "" {
		fmt.Fprintf(os.Stderr, "mt-webfetch: defuddle returned %d / empty body for %s\n", res.StatusCode, target)
		os.Exit(1)
	}
	os.Stdout.Write(body)
}
